File: csharp/error-handling/ErrorHandling.cs
using System;

public static class ErrorHandling
{
    public static void HandleErrorByThrowingException() => throw new Exception();

    public static int? HandleErrorByReturningNullableType(string input) => int.TryParse(input, out var value) ? value : null;

    public static bool HandleErrorWithOutParam(string input, out int result)
    {
        var x = HandleErrorByReturningNullableType(input);
        result = x ?? 0;
        return x != null;
    }

    public static void DisposableResourcesAreDisposedWhenExceptionIsThrown(IDisposable disposableObject)
    {
        disposableObject.Dispose();
        throw new Exception();
    }
}

File: fsharp/error-handling/ErrorHandling.fs
module ErrorHandling

open System

let handleErrorByThrowingException () = failwith "I am learning how to fail"

let handleErrorByReturningOption (input: string) : int option =
    match Int32.TryParse(input) with
    | true, value -> Some value
    | false, _ -> None

let handleErrorByReturningResult (input: string) : Result<int, string> =
    match handleErrorByReturningOption input with
    | Some value -> Ok value
    | None -> Error "Could not convert input to integer"

let bind switchFunction twoTrackInput =
    match twoTrackInput with
    | Ok value -> switchFunction value
    | Error message -> Error message

let cleanupDisposablesWhenThrowingException (resource: IDisposable) =
    try
        failwith "I got better at failing"
    finally
        resource.Dispose()
